background/rust.tex

```diff
@@ -331,80 +415,87 @@ \subsubsection{Lifetimes}
-
 \subsection{Concurrency Model} % (fold)
 \label{ssub:concurrency_model}
 
-Rust has a concurrency model. Duh.
-\todo {We should mention this shortly. This is not something that is very relevant for our project, but it is relevant for rust itself.}
+Solving both memory safety and concurrency might sound like two entirely different problems, but the ownership system described in \autoref{sec:guaranteed_memory_safety} actually turned out to go a long way of being the solution of them both \cite{web:fearless_concurrency_with_rust}.
+While we do not utilize or focus on the concurrency of Rust in this project, it is still a very important feature for the language, so it will shortly describe parts of it in this section.
 
-% subsubsection concurrency_model (end)
+As already mentioned, Rust's ownership system helps us to reason about the liveliness of variables, and it helps us to catch errors like use-after-free, and data races by not allowing us to have multiple references to the same mutable data.
+This makes it possible to describe and utilize many powerful idioms and paradigms, and concurrency models that are based on these core features of the language.
+Thus, Rust's way of providing concurrency is made possible because the ownership system guarantees that it will be safe.
+Indeed, the core concurrency functionality in Rust is merely an abstraction that is defined in the standard library, is is not a feature in the language itself.
+This is an important distinction, because it allows for any number 3rd party concurrency libraries to evolve based on the same core principle of ownership, without being dependent on the concurrency idioms provided in the standard library.
 
 \subsection{Unsafe Code} % (fold)
 \label{ssub:unsafe_code}
 
-The strong type system along with lifetime inference and borrow-checking help Rust become a safe language.
-The language provides safe, zero-cost abstractions over the machine it is running on, but it is not always possible to guarantee full safety without sacrificing performance, or vice versa.
-Achieving good performance is often a synonym with direct memory access and pointer manipulation, which in itself is not considered to be safe operations.
-\todo{At least by Rust. Depends on who considers it.}
+Rust's strong type system and static guarantees about memory safety goes a long way of verifying the safety of programs, but for certain programs the restrictions set by the compiler might be too conservative.
+There exists many programs that are indeed safe, but still not possible for the compiler to verify.
+Examples of code that the compiler cannot verify is code that utilize \emph{raw pointers}, i.e. pointers to mutable data (like a C pointer), code that call directly \emph{into} external C libraries through Rust's \gls{ffi}, or code that can be called \emph{from} \gls{ffi}.
+
+The ability to make new abstractions that optimize away to essentially nothing is an important feature of Rust.
```

Comment:
Suggest to rephrase to `The ability to make abstractions that optimize away with no overhead is an important design goal of Rust, called Zero Cost Abstraction.`